### ISSUE TITLE
Protected extension types or proper "Value types"

### ISSUE THREAD

**Issue by lrhn** [dart-lang/language#1467]
The current specification of extension types suggests `protected extension type` declarations to introduce a type which:
* is a zero overhead abstraction,
* not related to the `on` type (only to `Object?` and `Never`),
* can only be entered using a constructor, and
* can be exited by casting (can always be cast to `Object?`).

It's not a real wrapper object, although there is an extra part of the proposal to introduce `box`/`unbox` getters to do explicit boxing and unboxing.

To me, this all sounds very much like an attempt to define a *value type*, meaning a type with no (guaranteed) identity for values, allowing automatic boxing and unboxing as necessary, but otherwise looking like a class. The `show`/`hide` clauses allow us to forward some methods to the underlying value, which is a feature we could use in other situations too.

So, if we instead allowed a declaration like:
```dart
struct Int implements int as _value { // `struct` declares a "value class"
  final int _value;
  Foo(int value) : _value = value;
  Iterable<int> operator(int to) sync* {
    for (var i = this; i < to; i++) yield i;
  }
}
```
The `struct` works the same as `class`, but introduces a value without identity. This allows the compiler to inline/unbox and re-box the object as it pleases. I'd probably want `identical` to either always answer `false`, or always do `identical` comparisons on the field values. Structs do not introduce an interface and cannot be subclassed, or something like that (lots of possible details).

The `implements type as field` automatically adds forwarders for all non-overridden members of `type` to `field.member`. (We can perhaps also allow `show/hide` after the `as field`, but it's a problem that `show`/`hide` uses a comma separated list, just like `implements`).
There are also other options for forwarding, but this can handle *bulk* forwarding.

Together, these two features would give us all the affordability of protected extension types, except guaranteed zero overhead. It's a *compiler optimization* to inline structs where possible, but something that should be fairly predictable (whenever the static type is the value type, we can safely unbox it).

In return, we get general value type which can contain more than one value (extension types are only on *one* value), which work safely with generic containers (by auto-boxing), and where boxing/autoboxing is built into the system, not something you have to do manually using `box`/`unbox`. This is safer than the skin-deep abstraction of an extension type can ever be.

The forwarding feature can also be used by classes, and for more than one type at a time. It's a request we've had before (like https://github.com/dart-lang/sdk/issues/31483).

**Comment by tatumizer** [comment]
I can think of more intuitive syntax:
```dart
wrapper Int on int as value show int implements x, y, z {
   // no constructor necessary!
   // the wrapped values should be referred to as value inside the wrapper
}
// wrapper over tuple
// we don't allow wrappers over 1-tuples - see explanation below
// we can show parameter a and hide b while reading components from tuple value, and can use  undescored names 
// to automatically hide some components. To hide the named ones, use explicit "hide" clause
wrapper class Foo on (int a, double b, {String? name}) as tuple hide someTupleMethod implements Comparable<Foo> {
   int compareTo(Foo other) {...}
}
// invocation
Int(5).sign;
Int(5).extraMethod;
var x=Int(5); // no need to box it here: zero-cost declaration
x.sign; // inherited
x.extraMethod; // added in the wrapper
print(x); // auto-boxing!
var foo=Foo(1, 1.5, name: "Z"); // wrapper over a tuple
```
Example with a constructor and extra stuff inside:
```dart
wrapper Bar on int as value implements MyInterface {
   String _a;
   Bar() {... a="Hello"}; 
   factory Bar.fromString()=> ...
   //... etc
}
var bar=Bar(5); // heap allocation cannot be avoided here
```

Comments:
- We have to disallow 1-tuple to avoid confusion with between 1-tuple and a single-value
- "as" clause declares the name under which the wrapped object can be accessed from within 
- "this" refers to the wrapper object, not to the wrapped value - for consistency with other classes
- I don't like the word "struct" because it's basically the same as "tuple". IMO, we need to put greater emphasis on the fact that the class "wraps" the object. "Struct" doesn't convey this meaning. "Wrapper" might not be the most beautiful word, but it's used informally everywhere (and maybe formally too, in some other language, but I can't remember in which one)
- there's a lot more to discuss

I don't know what not to like here. This design is pretty straightforward. Doesn't leave too much room for confusion IMO. Whenever possible, compiler can optimize this thing by avoiding boxing, but the possibility of zero-cost is not the primary goal - it's just a nice side effect of the design.

**Comment by lrhn** [comment]
If we ignore the syntax differences, the significant difference is that you declare the variable in the class part (outside the body), I declare it in the body and reference it in the class part:
```dart
// No need to write `wrapper`, the forwarding should work for all classes.
class Bar on Foo _foo, Baz _baz implements Qux {  
  ...
}
```
vs.
```
class Bar implements Foo as _foo, Baz as _baz, Qux { 
  Foo _foo;
  Baz _baz;
  ...
}
```
Yours is definitely shorter in some cases, but (IMO) harder to read. I can't just look at the top of the object declaration and see the state.
Also, it doesn't allow two interfaces to be implemented by the same object:
```dart
class Foo implements Bar as _b, Baz as _b, Qux {
  BarBaz _b;
  ...
}
```

What it doesn't allow is exposing members which are not part of an interface. 
Maybe that could be handled at the declaration:
```dart
   Bar bar show bar1, bar2;
```
would introduce forwarders for `bar.bar1` and `bar.bar2` in the surrounding class. Then writing (my syntax) `implements Bar as bar` would be equivalent to adding (effectively, even if we don't have that syntax) `show Bar.*` on the `bar` declaration.


As for auto-unboxing, that's a separate feature. I wouldn't call it `wrapper class` because it's not about wrapping. It's about not having (or at least not needing to have) a consistent identity. I used `struct` because that's what C# uses for non-`Object` types.

**Comment by tatumizer** [comment]
I totally misunderstood your proposal. :-)
The main reason for misunderstanding was that just a while back, we were debating tuples aka Records, and we used to characterize tuples as (immutable) value types. So my (erroneous) interpretation of your proposal was that are trying to introduce another variant of tuple with added methods and interfaces, but call it "struct" for some reason. I thought wrapping an object around a tuple could be conceptually simpler.
But now I see that your proposal was more about forwarding. The mechanism of forwarding, however, doesn't have an obvious connection to value types, so it's not clear how these two concepts got entangled.

**Comment by lrhn** [comment]
The idea is that, as an *alternative* to "protected extension types", we should do "value types" (which can be unboxed and boxed as needed, and do not promise identity) and "easy forwarding" (for the `show`/`hide` part of protected extension types). Two features, which can both be used independently in other situations too, and which together makes protected extension types unnecessary. (So, basically, solve a more general problem, and not the too-specific and not-entirely-possible idea that zero-overhead abstraction can actually buy you any kind of safety).

**Comment by lrhn** [comment]
One problem with this idea is that value types are traditionally *unmodifiable* and *not subtypable*. If something does not have identity, mutating it makes little sense, and if you can replace it with something which has more state prevents unboxing.

For the non-subclassability, that probably means that you cannot extend (or implement) a `struct`. A `struct` can extend only classes (you can assign the struct to the supertype, it'll just get boxed). If you need to extend a struct, instead *contain* the struct and forward its members (which is another use for easy forwarding).

For unmodifiability, either we can just go with that (`struct` classes must have only final variables and extend only classes with `const` constructors, which luckily includes `Object`), or introduce some notion of "modification" which acts consistently with assignment, but isn't really. Basically, where mutating is known to be the same as replacing with a different struct. In that case, we don't want to be able to cast the struct to a mutable supertype, so again it can only extend classes with `const` constructrs, or maybe just `Object` (favor composition over inheritance, and all).

Example: `struct Foo {int i;} .... Foo x; x.i = 42;`, only allowed where `x` is itself a mutable reference to the struct, and it's equivalent to creating a new `Foo` with the updated value, and assigning that to `x`. Basically a `lhs.foo=42` is really a composite operation like <code>lhs = lhs[foo &mapsto; 42]</code> (where `e` is only evaluated once, and `lhs` must be an assignable expression). Can even do `lhs..foo = 1..bar = 2;` and incrementally update the `lhs` value. That should be optimizable in a lot of cases, but that means that we must prevent mutation of boxed structs, which means no `dynamic` assignments

Maybe just define `structValue..id = value` as evaluating to the *updated* struct. Then you have to explicitly write `x = x..foo = 42;` (read: `x` is now `x` where `foo` was changed to 42). Using `x.foo=42` as shorthand is an option, but less explicit.

If we introduce a supertype of all values, `Value`, which is a proper supertype of `Object` and `dynamic`, and where struct types are subtypes of `Value` (like in C#), then we also avoid such things as nullable structs and dynamic structs, those operations are only for (reference-)objects. Maybe we can make `void` be at the level of `Value`. It shouldn't be assignable to any other type anyway (and if it's still assignable to `dynamic`, its as an implicit downcast which can fail).

**Comment by tatumizer** [comment]
I think when we are talking about "value types", the real driving force is the desire to define a kind of "objects" to be allocated on stack. (Please reflect on your motivations to see if this is the case). Interestingly, this whole "stack" thing rarely enters the narrative - it stays in the shadows, while still controlling the plot. What I don't quite understand yet: why are we talking about those stack-allocated objects as if it's a given that we can easily implement them - and at the same time, the discussion of fat pointers provokes the controversy. But if on-stack allocation is so easy, let's forget about bits - just allocate 2 words for a fat pointer. After all, if the compiler knows how to manage on-stack structs in general, the same must be "even more true" for 2-word wrapper objects, isn't it? So what makes one of these things harder than another?

This is my cognitive dissonance *#1*. I also have *#2*: everyone is talking about the separation of the concept from the implementation. But if we discard the "small" implementation detail like "on-stack allocation" - why do we need value types at all? It's difficult for them to coexist with normal objects - there're all kinds of problems of boxing, casting, toString, hashCode, you name it. Maybe we don't need value types at all then? Imagine for a second that heap allocation is free. Will you still want value types?

When I'm trying to abstract from dissonances *#1* and *#2*, I think the main takeaway from the latest discussion (including this proposal) is that we can implement composition by delegation easily - using show/hide clauses (BTW, do you know any language where they are implemented? To me, this thing looks novel and very powerful - especially if we support a set of non-trivial, Regex-like, selectors to say what we want to show or hide), plus the way to delegate to more than one component (using `show x.someMethod, y.anotherMethod`). Do we really need to entangle "value types", "on-stack allocations" with this concept? 

And what if we always allocate on heap? Will the idea become worthless because of this?

**Comment by tatumizer** [comment]
A separate comment on another aspect of your post:

> Example: struct Foo {int i;} .... Foo x; x.i = 42; ... and it's equivalent to creating a new Foo with the updated value, and assigning that to x

I think this is not the best possible syntax. Less confusing syntax (IMO) would be the one we discussed before: `x = x.with(i: 42);`.
What's interesting about this syntax: it dictates the syntax of the constructor of Foo. Namely, no positional parameters - just the named ones (otherwise we will start arguing about the "names" of nameless parameters: is it `$1`? Not good. Maybe `[1]`? It sucks. Etc. :-))

**Comment by lrhn** [comment]
Stack allocation is a reason for value types, but it's only part of it. (Not just *stack* allocation, also inline-allocation in other objects, so let's call it *non-heap allocation*, where the "heap" here is the garbage collected object heap, not a C-like heap).

The reason for non-heap allocated objects is memory use reduction, memory churn reduction and therefore lower garbage collection overhead. It's space, and thereby time, optimization, not a semantic change in how the object works&mdash;at least not unless we need to add further *restrictions*.

So why do we talk about it at all? Because we actually do care about the overhead of allocating new objects. Maybe we shouldn't, and we should just optimize the normal objects better.
It's just that if we remove a "small" thing like identity, we know the compiler will have much more opportunities to optimize. That's why we're considering providing that feature. Just as when we consider allowing you to seal a class, it's because it allows the compiler to do more optimizations, and more *predictable* optimizations, than without it. 

The real goal here is good and *predictable* performance, because without that, Dart won't be a good platform for performance sensitive programs.

The cost of removing identity and allowing inline allocation is that you can't safely store a reference to a non-heap allocated object because either it might stop existing at any time (if it's on the stack) or it doesn't exist as an independent entity (if inlined in another object).
Not having a reference means that most `Object`-based operations won't work. You can't pass the value as an argument, which is value types will often have copy-semantics. You don't pass a reference, you create a new copy at the receiving side.
Then we get into design discussions about whether that copy itself has identity enough to allow mutation (so a value stored in a variable stays the same, but assigning it to another variable makes a copy), or whether it's all immutable and the compiler itself can completely forget about which value comes from where.

**Comment by tatumizer** [comment]
> Not just stack allocation, also inline-allocation in other objects, so let's call it non-heap allocation, where the "heap" here is the garbage collected object heap, not a C-like heap

Maybe "stack allocation" and "inline allocation" are two different things?
Consider a typical Flutter widget. The constructor has tons of parameters, each of them is an object requiring its own allocation, but many of those objects do not have any independent existence, they only make sense as parts of the host widget. 
But what prevents such objects from being inlined in the host widget? Inlined as Objects, with exactly the same layout as a normal Object? Then we can refer to this sub-object by reference, right?
In other words, when we call
```
var x=FooWidget( bar: Bar());
```
we can allocate Bar on stack temporarily, just to pass it to FooWidget constructor, which then copies (inlines) the thing into itself. Then we actually don't need structs, right? True, there are many details in there, it's just a raw idea. But have you tried to think along these lines?

**Comment by Cat-sushi** [comment]
I like value types with syntax of
```dart
struct Bar from Foo as _foo, Baz as _baz, Qux { 
  Foo _foo;
  Baz _baz;
  ...
}
```
without auto-forwarding by default,
because `ID` from `int` doesn't naturally have interface of int.

**Comment by tatumizer** [comment]
@lrhn: it occurred to me that in this and all earlier discussions of value types everybody overlooked a simple argument: value types, as soon as they are introduced, immediately imply that dart has to introduce pointers, too. Please check my logic:
- if we have struct as a value type, we will want to support lists of structs, too, where the list contains the unboxed structs as elements
- then I can write:
```dart
var list = <MyStruct>[ struct1, struct2]; // all unboxed
list[0].field0=42;
```
- what is the meaning of `list[0]` in this expression? Doesn't it mean that the compiler *takes* `list[0]` and copies it into the stack before assigning field0? Then it's not what the user wants! The value 42 will never find its way into the struct stored in the list!
- one may say: no, the compiler is smart enough to implement this operation like the user wants, without copying the struct on the stack
- but this doesn't solve the problem:
```dart
var list = <MyStruct>[ struct1, struct2]; // all unboxed
list[0].field0=42;
list[0].field1=43;
// BUT here I'm getting tired of repeating the same thing, so I write:
var el1 = list[1];
el1.field0 = 0;
el1.field1 = 1;
```
which is a showstopper (note that it's a trivial example: in reality, the chain can be much longer) I don't know how to implement it without explicitly introducing pointers.
Please let me know where my mistake is :-)

**Comment by lrhn** [comment]
I agree that we'll need *something*, hopefully not *pointers*. 

It's not a given that a "value type" is mutable. It such an overloaded word, used to mean both "data type" (immutable values) and "structs" (potentially mutable like objects, but without a permanent identity), or things in-between. You're talking about the struct.

Reference variables would work, something like `ref el1 = list[1];` would be an alias for `list[1]` as a *left hand side* as well, and then we'll could also have `ref` parameters. Structs are identified by the assignable expression (left-hand-side) that they are accessed through, they only change identity on assignment (including parameter passing).
It's a powerful feature, but not fundamentally different from capturing a variable with closures like `() => x` and `(v) { x = v; }`. 
A possible desugaring of `ref` variables is a pair of such functions, or only the getter for a final ref variable, basically abstracting over getters and setters, and then the "getter" *also* allows struct field assignment through it (so it's a triple of functions, get/set/updateStruct).

There is one other approach we can take, one which doesn't require pointers: Functional Updates!
* Structs are immutable.
* Assignment to a struct field *creates a new struct* and assigns it back to the variable holding the old struct.

Basically syntax like `structVariable.field = 42` really means `structVariable = structVariable._replace(field: 42)` (then that expression evaluates to 42 anyway). Works for any valid assignable LHS, so `list[index].field = 42` means `list[index] = list[index].replace(field: 42)` (with the usual caveats about not revaluating `list` and `index` twice). It's a composite assignment, like `+=42`, only the operator-assignment is `.field=`.
A final struct typed variable would be immutable (so would a `const` variable, obviously).

You can have nested structs and do `structVar.nest.field = 42` meaning `structVar = structVar._replace(nest: structVar.nest._replace(field: 42))`. 

It would be up to the compiler to optimize such assignments into in-place updates, or combine things `structvar..field1 = 42..field2 = 37` into a single `replace`.

It's not *quite* mutable structs, but it simulates it adequately (there are things about evaluation order too, for cases like `foo.field = expressionChangingFoo` which should not throw away the change. And it's not clear how well it works with `dynamic` types. I'm sure there are other details to figure out.

That's the one non-reference/non-pointer approach I've seen to updating structs, no pointer exposed, no reference variables needed, but it does rely on the struct variable being assignable.

**Comment by tatumizer** [comment]
Even with immutable structs, we have complications. Consider
```dart
var list = <MyStruct>[ struct1, struct2]; // all unboxed
var sum=list.filter((s)=>s.name=='Foo').fold(0, (accum, s)=>accum+s.value);
```
What is 's' in `filter((s)=>...)` ? Clearly, the compiler wouldn't want to pass the whole struct by value per se (that is, by copying it into the stack). It will pass a reference. There's no way around it. And this reference is a pointer type (\*) behind the scenes. It might be possible to pretend there's no pointer types, but they are there. At some point, the language will feel it's counterproductive to hide pointers (once they are there anyway), and introduce pointers explicitly, which leads to "unsafe mode". It's just one thing leads to another. Right? Do you know any language that has value types (structs) that doesn't have pointers?

(\*) Probably, a fat pointer. The parameter has to carry type information somehow.

**Comment by dmvvilela** [comment]
Any updates on this? Forgive me if im saying nonsense, but im wondering what if we had structs from day one before flutter even started. Considering SwiftUI (flutter style for building UIs) uses structs, maybe flutter could have worked with such instead of relying on stateful x stateless widgets classes. Just a thought, anyone more knowledgeable than me to have some good insights on the subject?

**Comment by kroolzone** [comment]
If dart had structs, we might not have needed codegen and macros for data class. Atleast for the majority of the cases data class is used by average user. I don't feel comfortable with making data classes with macros but I suppose its not a big deal.

**Comment by mraleph** [comment]
> Whats the team's thoughts on javascript potentially adding them? https://github.com/tc39/proposal-structs

JavaScript structs are just fixed layout objects. They are not value types. Dart already has this - any class instance is fixed layout.
